File: src/Functional/Scala/Adapter.scala
trait TemperatureReader {
  def readCelsius(): Int
}

final class LegacyFahrenheitSensor {
  def readFahrenheit(): Int = 86
}

final class FahrenheitSensorAdapter(legacy: LegacyFahrenheitSensor) extends TemperatureReader {
  override def readCelsius(): Int =
    Math.round((legacy.readFahrenheit() - 32) * 5.0 / 9.0).toInt
}

object Adapter {
  def main(args: Array[String]): Unit = {
    val legacy = new LegacyFahrenheitSensor
    val reader: TemperatureReader = new FahrenheitSensorAdapter(legacy)
    println(s"legacy=${legacy.readFahrenheit()}F")
    println(s"adapted=${reader.readCelsius()}C")
  }
}

File: src/Functional/Haskell/Adapter.hs
module Main where

data LegacyFahrenheitSensor = LegacyFahrenheitSensor

readFahrenheit :: LegacyFahrenheitSensor -> Int
readFahrenheit _ = 86

newtype TemperatureReader = TemperatureReader { readCelsius :: Int }

adaptTemperature :: LegacyFahrenheitSensor -> TemperatureReader
adaptTemperature sensor =
  let fahrenheit = readFahrenheit sensor
      celsius = round (fromIntegral (fahrenheit - 32) * 5 / 9 :: Double)
  in TemperatureReader celsius

main :: IO ()
main = do
  let legacy = LegacyFahrenheitSensor
      reader = adaptTemperature legacy
  putStrLn $ "legacy=" ++ show (readFahrenheit legacy) ++ "F"
  putStrLn $ "adapted=" ++ show (readCelsius reader) ++ "C"
